lib/os/onoff.c

```diff
@@ -365,23 +407,24 @@ int onoff_release(struct onoff_service *srv,
 		 * wait
 		 */
 		if ((k_is_in_isr() || k_is_pre_kernel())
-		    && ((srv->flags & ONOFF_SERVICE_STOP_SLEEPS) != 0)) {
+		    && ((mgr->flags & ONOFF_STOP_SLEEPS) != 0)) {
 			rv = -EWOULDBLOCK;
 			break;
 		}
 
 		stop = true;
 
-		set_service_state(srv, SERVICE_STATE_TO_OFF);
-		srv->releaser = cli;
+		set_service_state(mgr, ONOFF_STATE_TO_OFF);
+		set_client_type(cli, ONOFF_CLIENT_RELEASE);
+		sys_slist_append(&mgr->clients, &cli->node);
```

Comment:
since we have monitors now, then last releaser does not need to be enqueued to notify about stopping result. If that is applied then canceling simplifies a lot since release and request are instantly cancelled. There is probably no need to have client type field for the release. Also that would mean that `onoff_release` could be synchronous and actually anonymous (so also no release cancel, `onoff_cancel` -> `onoff_cancel_request`). Imo, that would be more natural for the client which care when service is on but does not care what happens when service is released. Same applies to monitor, which also seems more natural: one (or more) maintainer of the service which tries to help (applies reset) when something goes wrong. In that approach, `onoff_reset` could also be anonymous as monitor gets notified when service is restored.

Reply:
That's a pretty significant architectural change which suggest additional changes. But it has some potential; I'll think about this and look into it sometime next week.